### PR TITLE
E2E test improvements

### PR DESCRIPTION
**Pull request by grigasp**
Closes https://github.com/iTwin/viewer-components-react/issues/1048.

- Do not run e2e tests at all unless test-viewer, property-grid or tree-widget changes.
- If only tree-widget or property-widget changes, run only their tests.
- Build a production build of test-viewer for e2e tests.
- Create a docker container only for running tests and use pre-built test viewer there.
- Change the pipeline to GitHub Action and parallelize property-grid and test-viewer test runs.

TODO:
- Look into enabling [sharding](https://playwright.dev/docs/test-sharding). Might be complicated due to our use of docker.
- Some tests use transition-kind-of-actions like widget resize, etc. Those transitions may take up to several seconds, which add up and make tests slow. Could have a look at making those transitions quicker.

### CODE REVIEW

File: packages/itwin/property-grid/playwright.config.ts [hunk 1]
```diff
@@ -2,13 +2,28 @@
...
+
+import dotenv from "dotenv";
+import * as fs from "fs";
 import { defineConfig, devices } from "@playwright/test";
 
 /**
- * Read environment variables from file.
- * https://github.com/motdotla/dotenv
+ * Read auth client id and user credentials from `.env.e2e` file.
+ * The auth client must support scopes required by the viewer: "imodelaccess:read imodels:read realitydata:read".
```
[the remark this answers is not in the document]
Comment (grigasp) [reply packages/itwin/property-grid/playwright.config.ts]: Completely removed the mentioning of the three deprecated scopes, left only `itwin-platform`.

File: e2e.Dockerfile [hunk 2]
```diff
@@ -2,15 +2,28 @@
 # Copyright (c) Bentley Systems, Incorporated. All rights reserved.
 # Licensed under the MIT License. See LICENSE.md in the project root for license terms.
 #----------------------------------------------------------------------------------------------
-FROM viewer-components-react/test-viewer
+FROM mcr.microsoft.com/playwright:v1.48.2-jammy
```
[the remark this answers is not in the document]
Comment (grigasp) [reply e2e.Dockerfile]: Updated to `1.49.0-noble`

File: packages/itwin/property-grid/src/e2e-tests/auth.ts [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+import { getTestAccessToken } from "@itwin/oidc-signin-tool";
+import { test as setup } from "@playwright/test";
+
+setup("authenticate", async ({ page }) => {
+  if (!process.env.IMJS_USER_EMAIL || !process.env.IMJS_USER_PASSWORD) {
+    throw new Error(`User credentials need to be supplied through IMJS_USER_EMAIL and IMJS_USER_PASSWORD environment variables.`);
+  }
+  if (!process.env.IMJS_AUTH_CLIENT_CLIENT_ID) {
+    throw new Error(`Auth client id needs to be supplied through IMJS_AUTH_CLIENT_CLIENT_ID environment variable.`);
+  }
```
[the remark this answers is not in the document]
Comment (grigasp) [reply packages/itwin/property-grid/src/e2e-tests/auth.ts:16]: Setting it up was a requirement before these changes too, but you're right that we should have that. Updated the markdown.

**Comment by grigasp** [comment]
The e2e tests used to take ~1 hour, now they take ~15 min. - I think that warrants merging this PR asap and looking into other optimizations at some point in the future.

File: packages/itwin/tree-widget/src/e2e-tests/auth.ts [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+      password: process.env.IMJS_USER_PASSWORD,
+    },
+  );
+  await page.context().addCookies([
+    {
+      name: "IMJS_USER_ACCESS_TOKEN",
+      value: accessToken ?? "",
+      domain: "localhost",
+      path: "/",
+    },
```
Comment (MartynasStrazdas) [inline packages/itwin/tree-widget/src/e2e-tests/auth.ts:36]: Do cookies get cleaned before each e2e test run?

File: packages/itwin/tree-widget/src/e2e-tests/auth.ts [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+    throw new Error(`User credentials need to be supplied through IMJS_USER_EMAIL and IMJS_USER_PASSWORD environment variables.`);
+  }
+  if (!process.env.IMJS_AUTH_CLIENT_CLIENT_ID) {
+    throw new Error(`Auth client id needs to be supplied through IMJS_AUTH_CLIENT_CLIENT_ID environment variable.`);
+  }
+  // eslint-disable-next-line @itwin/no-internal
+  const accessToken = await getTestAccessToken(
+    {
+      authority: "https://qa-ims.bentley.com",
+      scope: "imodelaccess:read imodels:read realitydata:read",
```
[the remark this answers is not in the document]
Comment (grigasp) [reply packages/itwin/tree-widget/src/e2e-tests/auth.ts]: Yes, missed this one. Changed.

File: packages/itwin/tree-widget/src/e2e-tests/auth.ts (hunk 4 above)
Comment (grigasp) [reply packages/itwin/tree-widget/src/e2e-tests/auth.ts:36]: No. The `auth` script puts this value into a cookie and saves the browser context to a file. Then, before running the actual tests, we initialize the browser state from the saved file - this makes the cookie available to all the tests.

**Review by MartynasStrazdas — APPROVED** [review]

**Review by JonasDov — APPROVED** [review]

**Pull request merged**